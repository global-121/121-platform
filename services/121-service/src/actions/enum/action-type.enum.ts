export enum ActionType {
  startBlockNewPayment = 'start-block-new-payment',
  endBlockNewPayment = 'end-block-new-payment',
}
