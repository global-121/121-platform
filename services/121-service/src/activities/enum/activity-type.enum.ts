export enum ActivityTypeEnum {
  DataChange = 'data-change',
  FinancialServiceProviderChange = 'financial-service-provider-change',
  Message = 'message',
  Note = 'note',
  StatusChange = 'status-change',
  Transaction = 'transaction',
}
