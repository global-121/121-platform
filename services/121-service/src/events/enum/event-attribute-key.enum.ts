export enum EventAttributeKeyEnum {
  reason = 'reason',
  oldValue = 'oldValue',
  newValue = 'newValue',
  fieldName = 'fieldName',
  duplicateWithRegistrationId = 'duplicateWithRegistrationId',
  duplicateWithRegistrationProgramId = 'duplicateWithRegistrationProgramId',
}
