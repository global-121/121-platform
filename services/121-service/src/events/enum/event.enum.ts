export enum EventEnum {
  registrationDataChange = 'registrationDataChange',
  financialServiceProviderChange = 'financialServiceProviderChange',
  registrationStatusChange = 'registrationStatusChange',
  ignoredDuplicate = 'ignoredDuplicate',
}
