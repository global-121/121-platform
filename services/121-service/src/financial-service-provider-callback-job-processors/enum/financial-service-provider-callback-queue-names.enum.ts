export enum FinancialServiceProviderCallbackQueueNames {
  safaricomTransferCallback = 'safaricomTransferCallback',
  safaricomTimeoutCallback = 'safaricomTimeoutCallback',
}
