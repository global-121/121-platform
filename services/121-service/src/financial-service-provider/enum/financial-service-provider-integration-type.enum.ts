export enum FinancialServiceProviderIntegrationType {
  api = 'api',
  csv = 'csv',
  xml = 'xml',
}
