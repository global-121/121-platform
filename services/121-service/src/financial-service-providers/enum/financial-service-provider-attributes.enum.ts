export enum FinancialServiceProviderAttributes {
  phoneNumber = 'phoneNumber',
  nationalId = 'nationalId',
  fullName = 'fullName',
  addressStreet = 'addressStreet',
  addressHouseNumber = 'addressHouseNumber',
  addressHouseNumberAddition = 'addressHouseNumberAddition',
  addressPostalCode = 'addressPostalCode',
  addressCity = 'addressCity',
  bankAccountNumber = 'bankAccountNumber',
  whatsappPhoneNumber = 'whatsappPhoneNumber',
}
