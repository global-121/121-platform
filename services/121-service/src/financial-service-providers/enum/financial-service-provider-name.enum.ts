export enum FinancialServiceProviderName {
  intersolveVoucherWhatsapp = 'Intersolve-voucher-whatsapp',
  intersolveVoucherPaper = 'Intersolve-voucher-paper',
  intersolveVisa = 'Intersolve-visa',
  intersolveJumboPhysical = 'Intersolve-jumbo-physical',
  africasTalking = 'Africas-talking',
  belcash = 'BelCash',
  vodacash = 'VodaCash',
  bobFinance = 'BoB-finance',
  ukrPoshta = 'UkrPoshta',
  safaricom = 'Safaricom',
  commercialBankEthiopia = 'Commercial-bank-ethiopia',
  excel = 'Excel',
}

export enum FinancialServiceProviderConfigurationEnum {
  password = 'password',
  username = 'username',
  columnsToExport = 'columnsToExport',
  columnToMatch = 'columnToMatch',
  brandCode = 'brandCode',
  displayName = 'displayName',
  coverLetterCode = 'coverLetterCode',
}

export const FinancialServiceProviderConfigurationMapping: {
  [key in FinancialServiceProviderName]?: any;
} = {
  [FinancialServiceProviderName.intersolveVoucherWhatsapp]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveVoucherPaper]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveVisa]: [
    FinancialServiceProviderConfigurationEnum.brandCode,
    FinancialServiceProviderConfigurationEnum.coverLetterCode,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveJumboPhysical]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.africasTalking]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.belcash]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.vodacash]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.bobFinance]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.ukrPoshta]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.safaricom]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.commercialBankEthiopia]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.excel]: [
    FinancialServiceProviderConfigurationEnum.columnsToExport,
    FinancialServiceProviderConfigurationEnum.columnToMatch,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
};
