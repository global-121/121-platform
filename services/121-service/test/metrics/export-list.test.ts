import { RegistrationStatusEnum } from '@121-service/src/registration/enum/registration-status.enum';
import { RegistrationEntity } from '@121-service/src/registration/registration.entity';
import { DebugScope } from '@121-service/src/scripts/enum/debug-scope.enum';
import { SeedScript } from '@121-service/src/scripts/seed-script.enum';
import {
  registrationScopedGoesPv,
  registrationScopedMiddelburgPv,
  registrationsPV,
} from '@121-service/test/fixtures/scoped-registrations';
import {
  awaitChangePaStatus,
  deleteRegistrations,
  importRegistrations,
} from '@121-service/test/helpers/registration.helper';
import {
  getAccessToken,
  getAccessTokenScoped,
  getServer,
  resetDB,
} from '@121-service/test/helpers/utility.helper';
import {
  programIdOCW,
  programIdPV,
  registrationsOCW,
} from '@121-service/test/registrations/pagination/pagination-data';
import { HttpStatus } from '@nestjs/common';

function createExportObject(
  registration: Partial<RegistrationEntity> | any,
): RegistrationEntity | any {
  const exportObject = {
    ...registration,
  };
  delete exportObject.fspName;
  // remove empty values
  Object.keys(exportObject).forEach(
    (key) => !exportObject[key] && delete exportObject[key],
  );
  exportObject.paymentCount = 0;
  return exportObject;
}

describe('Metric export list', () => {
  const OcwProgramId = programIdOCW;
  const PvProgramId = programIdPV;
  let accessToken: string;

  beforeAll(async () => {
    await resetDB(SeedScript.nlrcMultiple);

    accessToken = await getAccessToken();
    await importRegistrations(OcwProgramId, registrationsOCW, accessToken);
    await deleteRegistrations(
      OcwProgramId,
      [registrationsOCW[0].referenceId],
      accessToken,
    );
    await awaitChangePaStatus(
      OcwProgramId,
      [registrationsOCW[1].referenceId],
      RegistrationStatusEnum.included,
      accessToken,
    );

    await importRegistrations(PvProgramId, registrationsPV, accessToken);
    await awaitChangePaStatus(
      PvProgramId,
      [registrationScopedMiddelburgPv.referenceId],
      RegistrationStatusEnum.included,
      accessToken,
    );
  });

  it('should export all people affected of a single program regardless of status', async () => {
    // Act
    const getRegistrationsResponse = await getServer()
      .get(`/programs/${OcwProgramId}/metrics/export-list/all-people-affected`)
      .set('Cookie', [accessToken])
      .send();

    // Assert
    const data = getRegistrationsResponse.body.data;
    expect(getRegistrationsResponse.status).toBe(HttpStatus.OK);
    expect(data.length).toBe(4);

    const expectedReferenceIds = registrationsOCW.map((r) => r.referenceId);

    // Also check if the right referenceIds are in the eport
    expect(data.map((r) => r.referenceId).sort()).toEqual(
      expectedReferenceIds.sort(),
    );

    for (const registration of registrationsOCW.slice(1)) {
      const exportRegistrationFound = data.find(
        (r) => r.referenceId === registration.referenceId,
      );
      expect(exportRegistrationFound).toMatchObject(
        createExportObject(registration),
      );
    }
  });

  it('should return all filtered registrations from 1 program using a filter for included and a scoped user', async () => {
    // Arrange
    const testScope = DebugScope.Zeeland;
    accessToken = await getAccessTokenScoped(testScope);

    // Act
    // 8 registrations in total are registered
    // 4 registrations are in include in program PV
    // 2 registrations of program PV and are in the scope (Zeeland) of the requesting user
    // 1 of those 2 registrations has status registered
    const getRegistrationsResponse = await getServer()
      .get(`/programs/${PvProgramId}/metrics/export-list/all-people-affected`)
      .set('Cookie', [accessToken])
      .query({
        ['filter.status']: `$ilike:registered`,
      })
      .send();

    // Assert
    const data = getRegistrationsResponse.body.data;
    expect(getRegistrationsResponse.status).toBe(HttpStatus.OK);
    expect(data.length).toBe(1);

    const exportRegistration = data[0];

    expect(exportRegistration).toMatchObject(
      createExportObject(registrationScopedGoesPv),
    );
  });

  it('should return all filtered registrations from 1 program using a filter and search query', async () => {
    // Arrange
    accessToken = await getAccessToken(); // gets admin access token

    // Act
    // 8 registrations in total are registered
    // 4 registrations are in include in program PV
    // 2 registrations of program PV have an attribute that contains '011' (phonenumber)
    // 1 of those 2 registrations has status registered
    const getRegistrationsResponse = await getServer()
      .get(`/programs/${PvProgramId}/metrics/export-list/all-people-affected`)
      .set('Cookie', [accessToken])
      .query({
        ['filter.status']: `$ilike:registered`,
        search: `011`,
      })
      .send();

    // Assert
    const data = getRegistrationsResponse.body.data;
    expect(getRegistrationsResponse.status).toBe(HttpStatus.OK);
    expect(data.length).toBe(1);

    const exportRegistration = data[0];

    expect(exportRegistration).toMatchObject(
      createExportObject(registrationScopedGoesPv),
    );
  });

  it('should export in excel format', async () => {
    // Arrange
    const testScope = DebugScope.Zeeland;
    accessToken = await getAccessTokenScoped(testScope);

    const getRegistrationsResponse = await getServer()
      .get(`/programs/${PvProgramId}/metrics/export-list/all-people-affected`)
      .set('Cookie', [accessToken])
      .responseType('blob')
      .query({
        format: 'xlsx',
      })
      .send();

    // Assert check if an excel is returned
    expect(getRegistrationsResponse.status).toBe(HttpStatus.OK);
    expect(getRegistrationsResponse.header['content-type']).toBe(
      'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
    );
    expect(Buffer.isBuffer(getRegistrationsResponse.body)).toBe(true);
  });
});
