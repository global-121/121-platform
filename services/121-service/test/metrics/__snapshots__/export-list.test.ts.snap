// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Metric export list should support using "select" to retrieve a specific subset of columns 1`] = `
[
  {
    "fullName": "Juliet Marsh",
    "phoneNumber": "15005550112",
    "referenceId": "reference-id-scoped-middelburg-pv",
  },
  {
    "fullName": "Jane Doe",
    "phoneNumber": "15005550111",
    "referenceId": "reference-id-scoped-goes-pv",
  },
  {
    "fullName": "Sam Winters",
    "phoneNumber": "15005550121",
    "referenceId": "reference-id-scoped-utrecht-pv",
  },
  {
    "fullName": "Nick Brouwers",
    "phoneNumber": "15005550200",
    "referenceId": "reference-id-not-scoped-pv",
  },
]
`;
